### ISSUE TITLE
react-native-svg

### ISSUE THREAD

**Issue by davidgithub1980** [expo/expo#2415]
Hi

Expo v30 includes react-native-svg **6.2.2**. Is there a reason why it does not support ^**7.0.0**? The not so new version significantly improves all its tools for the Android-powered devices.

**Comment by sjchmiela** [comment]
Hey! I've tried updating react-native-svg to 7.1.2 just now and it looks like it still has some issues, eg.

> Error while updating property 'height' in shadow node of type: RNSVGUse
> Value for height cannot be cast from Double to String

We're having an internal discussion whether to update RNSVG to 6.5.2 in SDK31 and then update to `^7.0.0` in SDK32. This would also give RNSVG time to stabilize API and internal changes. 🙂

**Comment by tanx** [comment]
SVG files exported via Sketch use the `Mask` element as well now. This doesn't work in the current release. We've had to go back to using PNG for now. But it looks like `react-native-svg@8.0.3` added support for `Mask`. So +1 for upgrading :)

**Comment by tanx** [comment]
> We're having an internal discussion whether to update RNSVG to 6.5.2 in SDK31 and then update to `^7.0.0` in SDK32. This would also give RNSVG time to stabilize API and internal changes. 🙂

If I may comment on this discussion, we're currently using `svgs` (which is compatible with `react-native-svg`) together with `react-native-web` for our desktop/electron app. It's nice to have a compatible solution across platforms within the react-native ecosystem and switching to RNSVG in expo might complicate that for this use case. Just my 2cents :) Thanks for considering it.

**Comment by sjchmiela** [comment]
I'll try the `8.0.3` release and see whether it's stable and works ok. 🙂

**Comment by tanx** [comment]
Cool thank you!

**Comment by jaimebiernaski** [comment]
Hi, @sjchmiela 

When trying the `8.0.3`, please check if `toDataURL()` works on Android.

I'm using Expo and react-native-svg, but  when I try  to send a SVG as png,

 ```
this.svg.toDataURL(base64 => {
        this._callback(base64);
      });
```
It crashes only on Android with the error:

> TypeError: undefined is not an object (evaluating 'RNSVGSvgViewManager.toDataURL')

On iOS, it works perfectly!

"expo": "^30.0.0",
"react-native": "https://github.com/expo/react-native/archive/sdk-30.0.0.tar.gz",

**Comment by msand** [comment]
@sjchmiela Great to see you're in the process of upgrading. If you find any regressions or issue please tell me and lets try to resolve them quickly so you can include the latest possible version. (I'm the current maintainer of react-native-svg)

**Comment by msand** [comment]
I just realised that the support for react-native style transform arrays (new feature) probably has a bad interaction with the hit testing (shouldn't be breaking old functionality), or the additional transform simply isn't considered in the hit testing. The matrix (if it's not null) should be inverted (in the view manager/property setter) and the touch points should be transformed using it (in the hit testing logic).

**Comment by msand** [comment]
@sjchmiela I've fixed a svg path spec conformance bug (old) in v8.0.4, and the hit testing when using rn-style transform arrays in v8.0.5
Would be nice to get these included, as otherwise I'll probably be getting various issues opened for problems with people stuck on an old version integrated into Expo (tends to take some time to find out people are using it, as most people don't include a repro or enough info), or people troubleshooting strange problems without knowing where to look.

**Comment by sjchmiela** [comment]
Great, thanks a lot! The only problem I encountered is that a clipped image isn't being rendered at all. I fixed the issue by doing:

```diff
--- a/android/src/main/java/com/horcrux/svg/VirtualView.java
+++ b/android/src/main/java/com/horcrux/svg/VirtualView.java
@@ -275,7 +275,7 @@ abstract public class VirtualView extends ViewGroup {
             ClipPathView mClipNode = (ClipPathView) getSvgView().getDefinedClipPath(mClipPath);
 
             if (mClipNode != null) {
-                Path clipPath = mClipNode.getPath(canvas, paint, Region.Op.UNION);
+                Path clipPath = mClipNode.getPath(canvas, paint);
                 switch (mClipRule) {
                     case CLIP_RULE_EVENODD:
                         clipPath.setFillType(Path.FillType.EVEN_ODD);
```
but I'm not sure that it really the proper solution.

I'm testing the behavior on Android 6.0.1. Good thing — Expo drops support for Android 4.4, so we don't really have to care here about Android KitKat support, and this is the commit that changed those lines recently — https://github.com/react-native-community/react-native-svg/commit/7bb1e748b7abe4822eeb68da245a0bba9d25bbb1.

**Comment by msand** [comment]
@sjchmiela The clipping implementation has changed to be spec conformant instead of the previous behaviour. Please check that the svg you're testing actually works as you expect inside modern browsers / other svg renderers as well. But, this might be a bug as well, hard to say without a repro.

**Comment by sjchmiela** [comment]
I was using the examples provided in react-native-svg repo, the clipping example icon showed a multicolor star without any cut inside and the two clipped image samples weren’t showing at all. 🤷‍♂️

**Comment by msand** [comment]
Oh, I'll have to test them, I suspect the multicolor star is likely not to be spec-conformant at least. The examples are very old and haven't been updated to account for improvements in spec-conformance.

**Comment by msand** [comment]
@sjchmiela Seems that change breaks spec conformance, in e.g. the example here: https://github.com/react-native-community/react-native-svg/issues/752#issuecomment-417819512
The star example seems to render according to spec in Android in the latest version, but not in iOS, there it still has the old behaviour for some reason. To see the expected result, check this fiddle: http://jsfiddle.net/f2q7c1y0/

**Comment by msand** [comment]
Then again, it seems setting clip-rule="evenodd" attribute on the clipPath rather than the G element which has the clip-path attribute, is broken in both.

**Comment by msand** [comment]
Fixed the regression and the spec conformance of clipPath and clip-rule: https://github.com/react-native-community/react-native-svg/commit/766926f75874dedd7b13dadf65315f7030c4234a
Released v8.0.6
<img width="1438" alt="screen shot 2018-10-24 at 22 11 19" src="https://user-images.githubusercontent.com/1131362/47455371-60b33600-d7da-11e8-815a-461708ceaccf.png">

**Comment by msand** [comment]
@sjchmiela Can you try with v8.0.6?

**Comment by msand** [comment]
The native code I tested was this:

```jsx
class ClipRule extends Component {
    static title = 'Clip a group with clipRule="evenodd"';
    render() {
        return (
            <Svg height={200} width={200}>
                <Defs>
                    <ClipPath id="c" clipRule="nonzero">
                        <Path d="M50 5L20 99l75-60H5l75 60z" />
                    </ClipPath>
                    <ClipPath id="b" clipRule="evenodd">
                        <Path d="M50 5L20 99l75-60H5l75 60z" />
                    </ClipPath>
                    <G id="a">
                        <Path fill="red" d="M0 0h50v50H0z" />
                        <Path fill="#00f" d="M50 0h50v50H50z" />
                        <Path fill="#ff0" d="M0 50h50v50H0z" />
                        <Path fill="green" d="M50 50h50v50H50z" />
                    </G>
                </Defs>
                <Use href="#a" clipPath="url(#b)" />
                <Use href="#a" clipPath="url(#c)" y={100} />
                <G transform="translate(100)">
                    <Use href="#a" clipPath="url(#b)" clipRule="evenodd" />
                    <Use href="#a" clipPath="url(#c)" y={100} />
                </G>
            </Svg>
        );
    }
}
```

**Comment by msand** [comment]
Published v8.0.7 with: [android] Support image onLayout, transform arrays, and fix clipping. https://github.com/react-native-community/react-native-svg/commit/0fb4177101d02043dd6fc160cc5c5cb7788266f6
@sjchmiela Were there any other issues you have found as of yet?

**Comment by sjchmiela** [comment]
I don’t think so, I’ve only used examples provided in the repo and some simple `victory-native` chart. Apart from this clipping issue everything looked pretty good. Thanks for all your work, especially all the recent fixes! 👏 As I’m logging off for today 😴, @esamelson will take on this task if he finds spare time, otherwise I’ll upgrade the library morning CEST. Again, thank you! 🙇‍♂️

**Comment by tanx** [comment]
Wow great work everyone. thank you!
